pkg/controller/predicate.go

```diff
@@ -0,0 +1,134 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// NewOwnedObjectPredicate creates a new predicate that filters only relevant owned object events,
+// such as creating or updating an object without an owner or with an owner that is being deleted,
+// or deleting an object with an owner that is not being deleted.
+func NewOwnedObjectPredicate(ownerType client.Object, reader client.Reader, ownerMapper Mapper, finalizer string, logger logr.Logger) predicate.Predicate {
+	return &ownedObjectPredicate{
+		ownerType:   ownerType,
+		reader:      reader,
+		ownerMapper: ownerMapper,
+		finalizer:   finalizer,
+		logger:      logger,
+	}
+}
+
+type ownedObjectPredicate struct {
+	reader      client.Reader
+	ownerType   client.Object
+	ownerMapper Mapper
+	finalizer   string
+	logger      logr.Logger
+}
+
+// Create returns true if the Create event should be processed.
+func (p *ownedObjectPredicate) Create(e event.CreateEvent) bool {
+	if e.Object == nil {
+		p.logger.Error(nil, "CreateEvent has no object", "event", e)
+		return false
+	}
+	return p.handleCreateOrUpdateEvent(e.Object)
+}
+
+// Update returns true if the Update event should be processed.
+func (p *ownedObjectPredicate) Update(e event.UpdateEvent) bool {
+	if e.ObjectNew == nil {
+		p.logger.Error(nil, "UpdateEvent has no new object", "event", e)
+		return false
+	}
+	return p.handleCreateOrUpdateEvent(e.ObjectNew)
+}
+
+// Delete returns true if the Delete event should be processed.
+func (p *ownedObjectPredicate) Delete(e event.DeleteEvent) bool {
+	if e.Object == nil {
+		p.logger.Error(nil, "DeleteEvent has no object", "event", e)
+		return false
+	}
+	return p.handleDeleteEvent(e.Object)
+}
+
+// Generic returns true if the Generic event should be processed.
+func (p *ownedObjectPredicate) Generic(event.GenericEvent) bool {
+	return false
+}
+
+func (p *ownedObjectPredicate) handleCreateOrUpdateEvent(obj client.Object) bool {
+	ownerKey := p.ownerMapper.Map(obj)
+	if ownerKey.Name == "" {
+		return false
+	}
+	logger := p.logger.WithValues("name", obj.GetName(), "namespace", obj.GetNamespace(), "ownerName", ownerKey.Name, "ownerNamespace", ownerKey.Namespace)
+	owner, err := p.getOwner(ownerKey)
+	if err != nil {
+		return false
+	}
+	if owner == nil {
+		logger.Info("Creating or updating an object without an owner")
+		return true
+	}
+	if !controllerutil.ContainsFinalizer(owner, p.finalizer) {
```

Comment:
Why are you exiting early here?

Reply:
Because an owner without a finalizer is not considered "valid" and the subsequent logic that compares the state of the object deletion to the one of owner deletion may be flawed in this case. For example, services that are not of type `LoadBalancer` would not be valid owners of `PublicIPAddresses`. I admit this is a very general explanation for the "create or update" case and I am not able to point to an actual problem here (it is possible for the "delete" case). On the other hand, if you don't see actual issues with having this `if` I would rather keep it for consistency.

Reply:
Ok, thanks for explaining.